datafusion/optimizer/src/propagate_empty_relation.rs

```diff
@@ -154,14 +156,14 @@ impl OptimizerRule for PropagateEmptyRelation {
                         Ok(Transformed::yes(LogicalPlan::Projection(
                             Projection::new_from_schema(
                                 Arc::new(child),
-                                plan.schema().clone(),
+                                input_schema.clone(),
                             ),
                         )))
                     }
                 } else {
                     Ok(Transformed::yes(LogicalPlan::Union(Union {
                         inputs: new_inputs,
-                        schema: union.schema.clone(),
+                        schema: input_schema.clone(),
```

Comment:
From the example, after `propagate_empty_relation`, the new inputs of `Union` doesn't have the Alias 'a', but only `test` exists.
If we want to use the new inputs's schema, it will be 
`DFSchema { inner: Schema { fields: [Field { name: "col_int32", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [Some(Bare { table: "test" })], functional_dependencies: FunctionalDependencies { deps: [] } }`

However, the schema before is `DFSchema { inner: Schema { fields: [Field { name: "col_int32", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [Some(Bare { table: "a" })], functional_dependencies: FunctionalDependencies { deps: [] } }`

The only difference is `field_qualifiers`, but `assert_schema_is_the_same` will throw an error if schema changed during the optimization.

Reply:
I think a solution may be, when creating the schema for a plan and for Alias{expr, name}, we use the expr's name rather than the `name`, however this would also be a big changing I think. 🥲

Reply:
I would have thought that Union required all the inputs to have the exact same schema (and be properly coerced / aliased to make that the case)

Maybe there is a bug when constructing Union somewhere that permits the inputs to have different schemas 🤔

Reply:
Or from earlier, what if when we build the schema for `Project` and the input for the `Project` is an `Alias`, we use Alias's expr name rather than Alias's name. 🤔  I'm not sure whether it would break some criteria.

Reply:
🤔  maybe we also need add an alias for columns in the subsequent children so the column names match exactly 🤔

Reply:
😢That's true. I'm kind of confusing how to deal with it.

Reply:
It looks like LogicalPlanBuilder::union has the code to do the schema coercion / matching

https://github.com/apache/datafusion/blob/405a5f60406f8e2757d2d6fb8e3cc46094aa8ed7/datafusion/expr/src/logical_plan/builder.rs#L1365-L1429

So perhaps we need to either:
1. Fix the `LogicalPlanBuilder` code
2. Update optimizer passes to not create `LogicalPlan::Union` directly but instead use the builder (and thus apply the schema rationalization to occur)

Reply:
Yes, I understand you.
But for solution 1, like I said in https://github.com/apache/datafusion/pull/10494#discussion_r1600567694, the merged schema would use the `left_qualifier` in line 1395 - 1396, i.e. the first LogicalPlan's.
```
left_qualifier.cloned(), 
Arc::new(Field::new(left_field.name(), data_type, nullable)),
```
However, I'm not quiet clear how to set the correct qualifier, they are different due to a `SubqueryAlias`🤔. Choosing either one would make some case failed. 😞